nixos/modules/virtualisation/containerd.nix

```diff
@@ -97,5 +133,68 @@ in
         StartLimitIntervalSec = "120s";
       };
     };
+
+    # User services adapted from
+    # https://github.com/containerd/nerdctl/blob/master/extras/rootless/containerd-rootless-setuptool.sh
+    systemd.user.services = lib.mkIf (cfg.nerdctl.enable) {
+      containerd = {
+        wantedBy = [ "default.target" ];
+
+        description = "containerd (Rootless)";
+        environment = {
+          CONTAINERD_ROOTLESS_ROOTLESSKIT_PORT_DRIVER = cfg.nerdctl.portDriver;
+        };
+        # needs newuidmap
+        path = [ "/run/wrappers" ]
+          ++ lib.optional config.boot.zfs.enabled config.boot.zfs.package;
+        unitConfig = {
+          ConditionUser = "!root";
+          StartLimitInterval = "60s";
+        };
+
+        serviceConfig = {
+          ExecStart = "${pkgs.nerdctl}/bin/containerd-rootless";
+          ExecReload = "${pkgs.procps}/bin/kill -s HUP $MAINPID";
+          TimeoutSec = 0;
+          RestartSec = 2;
+          Restart = "always";
+          StartLimitBurst = 3;
+          LimitNOFILE = "infinity";
+          LimitNPROC = "infinity";
+          LimitCORE = "infinity";
+          TasksMax = "infinity";
+          Delegate = true;
+          Type = "simple";
+          KillMode = "mixed";
+        };
+      };
+
+
+      buildkit = {
+        wantedBy = [ "default.target" ];
+
+        description = "BuildKit (Rootless)";
+        partOf = [ containerdUnit ];
+        path = with pkgs; [ buildkit coreutils util-linux ];
+        serviceConfig =
+          let
+            # Helper script:
+            # https://github.com/containerd/nerdctl/blob/884dc5480da0c4db5e2e18b008a9a7578af59b51/extras/rootless/containerd-rootless-setuptool.sh#L142-L147
+            nsenterScript = pkgs.writeShellScript "nsenter"
+              ''
+                pid=$(cat "$XDG_RUNTIME_DIR/containerd-rootless/child_pid")
+                exec nsenter --no-fork --wd="$(pwd)" --preserve-credentials -m -n -U -t "$pid" -- "$@"
+              '';
```

Comment:
I'd suggest to just use `containerd-rootless-setuptool.sh nsenter`

Reply:
I feel like that should live in `containerd-rootless.sh` since it's runtime rather than setup & the buildkit service you install doesn't use `containerd-rootless-setuptool.sh nsenter`
https://github.com/containerd/nerdctl/blob/main/extras/rootless/containerd-rootless-setuptool.sh#L265